### DISCUSSION TITLE
Options for testcard?

### DISCUSSION THREAD

**Discussion by LsrSal** [arkypita/LaserGRBL#1917] [General]
Running "test card" is essential to calibrate laser power and feed rates. My understanding that many commercial software had test grid generation capability built in. I'm on the hobby budget, meaning currently no budget for commercial software. With LaserGRBL I see no obvious way to effectively calibrate laser power.

I came to the point where I run grayscale grid as photo and then manually edited speed and power commands. Not very elective way to do that, especially if needed grid for very specific narrow ranges.
Then I've started coding my own app: in browser pure JS calibration grid generator. While working on it I wonder how most users of LaserGRBL resolve their need to calibrate?

Is there an easy way? I do not see many references to it.

I think addition of such tool is essential for such fantastic piece of software as LaserGRBL.

For now I'm still practicing very cumbersome way to calibrate by:
-creating custom material
-loading my reference grayscale image
-running burn and tweaking override on the fly
-guessing better settings and creating another material and trying it again
-repeat.......

Any suggestions about best way to calibrate laser with LaserGRBL?

**Comment by Ams-Laser** [comment]
It would be a nice feature
Btw, are you considering sharing your code?

**Reply by LsrSal** [reply]
Sure, if there is nothing else and my code, when completed, is reasonably functional - I will be happy if others will find it useful.

The question is about if there is a solution already present (that I've missed), or how crowd using LaserGRBL resolving this need for themselves.

**Reply by LsrSal** [reply]
I guess you are immediately interested. I would be curious to get some feedback.

**Comment by LsrSal** [comment]
I guess waiting does not bear much fruit here. So I did finish main part of the my code.

I hope my app is a good compliment to LaserGRBL. I hope some day to see similar functionality integrated with it. For now - you may conveniently generate grid to make calibration of you laser so much easier.

Main function is working. I will polish it when have time. Please feel free to participate and contribute.
Please check: [https://github.com/LsrSal/Laser_Test_Card_Generator](https://github.com/LsrSal/Laser_Test_Card_Generator)

**Reply by StuartB4** [reply]
Hi LsrSal. Your link actually goes to here. 

"https//github.com/arkypita/LaserGRBL/discussions/url"

This one is correct. 
"https://github.com/LsrSal/Laser_Test_Card_Generator"

The grid appears to be working fine though. Can't test it properly because my engraver is out of action at the minute. 

![image](https://user-images.githubusercontent.com/16980026/189336752-4491934b-478c-4ce8-9a74-a139e2923749.png)

**Reply by LsrSal** [reply]
Got it, thanks, link fixed.

**Reply by StuartB4** [reply]
👍🙂

**Comment by Tribo-2dot0** [comment]
Hello. Your application is fantastic, the LaserGrbl community is certainly grateful.
It would be possible to implement the choice of bi-directional engraving, and the choice of the direction, vertical or horizontal, of the engraving, some lasers have the diode oriented in the vertical direction (like my 10W Laser Tree).

**Reply by LsrSal** [reply]
Thanks!

I can add fill scan direction, no problem. Next revision.

Bidirectional fill: that would be for the fill/raster only. Most of the test card elements are vector, therefore dual direction is not applicable for the most part, only for the fill. My intent for the test card is to map and optimize power settings. There is little difference in relation to power/fill density if uni or bidirectional burn. Bidirectional useful to compensate system delay to avoid interlacing, but LaserGRBL does not implement such compensation anyway, so it will be of little use. For most quality work with LaserGRBL, for raster it must be used unidirectional anyway. I see no practical reason for bidirectional fill in the test card. I may do it just for fun anyway.

On second though it might be useful for tuning mechanics to reduce bounce/excitation/mechanical oscillation. I may think of other tests for this.

If you and anybody has ideas for other useful functions of test card generator - please stop by at my GitHUB and share your ideas.
I might have some time in a month or so to update my code with new functions.

I hope some day Arkupita will integrate it into LaserGRBL.

**Reply by Tribo-2dot0** [reply]
Bi-directional engraving saves time, the difference in final quality is practically imperceptible and compensating. LaserGrbl has the option to turn unidirectional engraving on. In test card the utility would be more to save time... but certainly the option of being able to choose between vertical and horizontal is the best choice between the two, as it refers directly to the dpi.
Once again congratulations for the work.
